### PR TITLE
optimization for x-repo search index builds

### PR DESCRIPTION
**Pull request by jctanner**
https://issues.redhat.com/browse/AAH-2337
https://issues.redhat.com/browse/AAH-2351
https://github.com/pulp/pulp_ansible/issues/1467

In the previous iteration of this code, ALL collection indexes would be rebuilt if at least one of their CVs were still in the repository version. This patch uses the .added and .removed functions to make a set of namespace(s).name(s) which have changed in the newest repository version. This set of namespace(s).names(s) can then be used to skip rebuilding the index for a collection which hasn't changed in some way.

Without the patch:
image

With the patch:
image

### CODE REVIEW

File: pulp_ansible/app/tasks/collectionversion_index.py [hunk 1]
```diff
@@ -93,17 +129,14 @@ def update_index(distribution=None, repository=None, repository_version=None, is
         if CVIndex.objects.filter(repository_version=repository_version).exists():
             return
 
+    # What has changed between this version and the last?
+    changed_collections = compute_repository_changes(repository_version)
+
     # get all CVs in this repository version
-    cvs = repository_version.content.filter(pulp_type="ansible.collection_version").values_list(
+    cvs_pk = repository_version.content.filter(pulp_type="ansible.collection_version").values_list(
```
Comment (mdellweg) [inline pulp_ansible/app/tasks/collectionversion_index.py:136]: I'm kind of used to this convention. (Feel free to ignore)
```suggestion
    cv_pks = repository_version.content.filter(pulp_type="ansible.collection_version").values_list(
```

File: pulp_ansible/app/tasks/collectionversion_index.py [hunk 2]
```diff
@@ -35,6 +37,40 @@ def get_highest_version_string_from_cv_objects(cv_objects):
...
+    ).first()
+
+    # If there isn't a previous verison, all things have "changed"
+    if previous_version is None:
+        return None
+
+    changed_collections = set()
+
+    for method in ["added", "removed"]:
+        func = getattr(repository_version, method)
```
Comment (mdellweg) [inline pulp_ansible/app/tasks/collectionversion_index.py:57]:
```suggestion
    for func in [repository_version.added, repository_version.removed]:
```

File: pulp_ansible/app/tasks/collectionversion_index.py [hunk 3]
```diff
@@ -35,6 +37,40 @@ def get_highest_version_string_from_cv_objects(cv_objects):
...
+        for modified in func(base_version=previous_version):
+            if modified.pulp_type == CollectionVersion.get_pulp_type():
+                cv = modified.ansible_collectionversion
+                changed_collections.add((cv.namespace, cv.name))
+            elif modified.pulp_type == AnsibleCollectionDeprecated.get_pulp_type():
+                dep = modified.ansible_ansiblecollectiondeprecated
```
Comment (mdellweg) [inline pulp_ansible/app/tasks/collectionversion_index.py:63]: I know, the scope is small, but "dep" is a bit ambiguous. (Sounds like dependency to me.)

File: pulp_ansible/app/tasks/collectionversion_index.py [hunk 4]
```diff
@@ -35,6 +37,40 @@ def get_highest_version_string_from_cv_objects(cv_objects):
...
+    # If there isn't a previous verison, all things have "changed"
+    if previous_version is None:
+        return None
+
+    changed_collections = set()
+
+    for method in ["added", "removed"]:
+        func = getattr(repository_version, method)
+        for modified in func(base_version=previous_version):
+            if modified.pulp_type == CollectionVersion.get_pulp_type():
```
Comment (mdellweg) [inline pulp_ansible/app/tasks/collectionversion_index.py:59]: Is it worth caching that pulp-type in a variable?

File: pulp_ansible/app/tasks/collectionversion_index.py [hunk 5]
```diff
@@ -35,6 +37,40 @@ def get_highest_version_string_from_cv_objects(cv_objects):
...
+    # Figure out what the previous repo version is
+    repository = repository_version.repository
+    previous_number = repository_version.number - 1
+    previous_version = RepositoryVersion.objects.filter(
+        repository=repository, number=previous_number
+    ).first()
+
+    # If there isn't a previous verison, all things have "changed"
+    if previous_version is None:
+        return None
```
Comment (mdellweg) [inline pulp_ansible/app/tasks/collectionversion_index.py:52]: Does this mean with `retain_repo_versions = 1` there is no optimization?

**Review by mdellweg — APPROVED** [review]
Some comments/questions, but nothing really blocking.

**Review by gerrod3 — APPROVED** [review]

**Pull request merged**